### ISSUE TITLE
[Tracking] Pi0.5 Integration and Runtime Optimization Roadmap

### ISSUE THREAD

**Issue by chenchaoxu7575** [vllm-project/vllm-omni#6867]
## Scope

This tracker turns [#6524](https://github.com/vllm-project/vllm-omni/issues/6524) into claimable tasks for Pi0.5 integration, Pi-family structure, model-runtime optimization, and scheduler-level batching. Each task states its type, dependency, estimated code size, hardware, implementation direction, and acceptance criteria.

Pi0.5 functionality is the baseline. Runtime work must keep a portable eager fallback and must not trade correctness for benchmark results. Post-training, MEM, Real-Time Chunking (RTC), timestep reduction, and sequence parallelism are not part of this milestone.

Difficulty estimates count implementation and test changes, excluding generated and lock files: **S** <=200 LOC, **M** 201-500, **L** 501-1000, **XL** >1000.

Status: **Ready** can be claimed now; **In progress** has an owner; **Blocked** waits for its start gate; **Design** can be explored before implementation/performance sign-off.

## Roadmap

```text
F1 ──> F2
 ├──> S0
 ├──> R0 ──> R1 ──> R2
 └──> D0 design/prototype
               ^
               └── R1 is required for D0 performance sign-off
```

The R1 interface study may start in parallel with R0, but its integration and performance decision use R0 as the optimized model-local baseline. D0 design can proceed after F1, while its performance result waits for R1. R2 later provides the optimized formed-batch comparator.

| ID | Type | Deliverable | Status | Start gate | Difficulty | Hardware | Owner / Link |
|---|---|---|---|---|---:|---|---|
| **F1** | Feature | Pi0.5 functional integration and tests | **In progress** | None | XL | Supported CUDA GPU | @chenchaoxu7575 / #6950 |
| **F2** | Framework adjustment | Pi-family shared model/runtime structure | **Blocked** | F1 | XL | Supported CUDA GPU | / |
| **S0** | Performance | Profile and reduce BS=1 engine overhead | **Blocked** | F1 | L | Data-center GPU; 4090/5090 baseline | / |
| **R0** | Performance | Clean up and rebase the #4419 BS=1 optimized runtime | **Blocked** | F1 | XL | Data-center GPU; 4090/5090 baseline | / |
| **R1** | Performance + Framework integration | Evaluate and integrate reusable DiT runtime components | **Design** | F1 study; R0 final evaluation | XL | Same as R0 | / |
| **R2** | Feature + Performance | Formed-batch Pi0.5 execution, B=1/2/4/8/16/32 | **Blocked** | R1 | XL | Data-center GPU; 80 GB recommended | / |
| **D0** | Parallel feature + Performance experiment | Scheduler-visible denoise steps and Continuous Batching | **Design** | F1 prototype; R1 perf | XL | Data-center GPU; 80 GB recommended | / |

## Common benchmark contract

- Reference: LeRobot v0.6.1, checkpoint `lerobot/pi05_base`, tokenizer `google/paligemma-3b-pt-224`.
- Workload: BF16, action horizon 50, 10 flow steps, 1/2/3 image views. Language input is always padded/truncated to 200 tokens; actual length is carried by the attention mask and is not a graph bucket.
- Performance tasks require an NVIDIA data-center CUDA GPU with sufficient memory; A100 and H100 are both valid. Use the same GPU model for each before/after comparison, and record the exact GPU, software stack, clocks/power policy, and memory capacity. An 80 GB card is recommended for the full B=32 matrix.
- For BS=1 performance tasks, also report 4090/5090 latency baselines. Consumer-card results are informative and are not mixed with data-center-GPU pass/fail comparisons.
- Report model-only, in-process pipeline, engine, and OpenPI E2E separately. Report warm p50/p95, throughput where applicable, cold compile/autotune/capture time, peak memory, software versions, and raw results.
- Correctness comparisons use identical inputs, checkpoint, timestep schedule, and fixed noise. Performance comparisons use the same process configuration and must exceed measured run-to-run variation.

### F1 — Pi0.5 functional integration

**Problem and deliverable.** Pi0.5 is not available in upstream vLLM-Omni. Add the model, config, processor, pipeline, registry/deploy/docs, constrained OpenPI per-request option forwarding, and correctness/serving tests. Relative actions are required. MEM and RTC are unsupported in this milestone and must fail explicitly when requested.

**Suggested approach.** Follow merged Pi0 [#4222](https://github.com/vllm-project/vllm-omni/pull/4222) for repository structure, LeRobot v0.6.1 for semantics, and closed PR [#4419](https://github.com/vllm-project/vllm-omni/pull/4419) only as an implementation reference. Keep CUDA Graph and fused kernels out of the functional baseline.

**Acceptance.** With `lerobot/pi05_base`, fixed noise, and 10 steps, 1/2/3-view outputs match LeRobot at `atol=1e-4`; preprocessing, normalization, relative/absolute action conversion, parameter forwarding, unit tests, and OpenPI E2E pass. Unknown or declared-but-unconsumed options are rejected rather than silently ignored. No performance gate.

### F2 — Pi-family shared structure

**Problem and deliverable.** Pi0 and Pi0.5 duplicate model modules, preprocessing/configuration, and pipeline integration. Produce a short Pi0/Pi0.5 difference table, then extract only genuinely shared Pi-family modules and runtime-facing contracts.

**Suggested approach.** Derive the boundary from the two working implementations: share identical vision-language/action-expert building blocks, prefix/mask helpers, flow-matching utilities, processor/config plumbing, and portable operator interfaces; keep model-specific semantics explicit. Do not introduce a universal `VLABaseModel`.

**Acceptance.** Pi0 and Pi0.5 unit, parity, and OpenPI serving tests remain unchanged; both eager paths remain directly usable; any model/pipeline BS=1 latency change is reported against measured run-to-run variation. The PR includes the difference table and documents which structures remain intentionally model-specific.

### S0 — Profile and reduce BS=1 engine overhead

**Problem and deliverable.** Existing data separates model, pipeline, engine, and OpenPI E2E, but does not attribute or reduce the engine-path overhead. Instrument request enqueue/queueing, orchestrator construction, stage submission, scheduler/executor dispatch, output routing, and response handling, then implement the highest-value low-risk reductions.

**Suggested approach.** Profile first, then evaluate a single-stage direct/inline path, removal of redundant tensor/array copies, and reuse of request/output allocations. Preserve the generic multi-stage path and use the four-layer methodology from the [#6524 profile](https://github.com/vllm-project/vllm-omni/issues/6524#issuecomment-5448068126).

**Acceptance.** Pi0.5 output and fallback behavior are unchanged. On the selected data-center GPU, report before/after warm p50/p95 for each engine segment and for `engine latency - pipeline latency`; any claimed improvement must exceed measured run-to-run variation, with no unexplained E2E or tail-latency regression. Report the same BS=1 breakdown on 4090/5090 and provide the harness, raw results, and attribution. If no tested change produces a reproducible improvement, document the result and close this optimization direction without landing extra complexity.

### R0 — Clean up and rebase the #4419 BS=1 optimized runtime

**Problem and deliverable.** PR #4419 mixes Pi0.5 functionality with a large model-local fast path and is based on an older repository state. Rebase only its useful BS=1 optimizations onto F1: CUDA Graph, selected Triton kernels, static buffers/workspaces, weight packing, and prefix/denoise fast paths. Remove duplicated functional code and expose the result as one explicit optimized backend with an eager fallback.

**Suggested approach.** Port #4419 component by component and keep an A/B for each retained optimization. Compare two graph topologies: (1) a preferred modular design with image-encoder, VLM-prefix, and single-denoise-step graph families, replaying the denoise graph 10 times; and (2) a monolithic whole-loop graph based on #4419's combined prefix/denoise approach. “Three graphs” means three graph boundaries/families; the number of captured objects is still bounded by the declared shape set. Prefer the modular design unless identical-workload A/B shows a reproducible latency regression; otherwise report graph-launch/buffer-handoff cost and retain the monolithic path as an optional backend. Cover 1/2/3 views and fixed 200-token language input; prewarm known shapes and fall back instead of building synchronously on an online request. R0 may remain model-local: framework reuse is evaluated in R1.

**Acceptance.** The optimized and eager paths have fixed-noise parity for 1/2/3 views and short/normal/near-limit prompts. Graph count and memory are bounded by configuration rather than request history. On at least one data-center GPU, the retained backend shows a reproducible warm model-latency improvement over F1 eager beyond measured run-to-run variation. Report per-component benefit, cold compile/autotune/capture cost, peak memory, and 4090/5090 BS=1 baselines.

### R1 — Evaluate and integrate reusable DiT runtime components

**Problem and deliverable.** R0 provides a correct model-local optimized baseline, but it may duplicate capabilities already present in the vLLM-Omni diffusion runtime. Determine which R0 mechanisms can reuse existing compile, attention backend, `CustomOp`, cache, graph/runtime, and step-execution interfaces, then integrate the parts that have a clear fit.

**Suggested approach.** Produce a reuse matrix with one decision per R0 component: **Adopt existing runtime**, **Keep model-local**, or **Needs framework change**. Validate the important decisions with prototypes and identical-input A/B results; see [Sub-RFC 2](https://github.com/vllm-project/vllm-omni/issues/6524#issuecomment-5404423504). Do not force exact prefix KV state into an unrelated approximate diffusion cache abstraction, and do not introduce a universal VLA backend.

**Acceptance.** The reuse matrix covers every R0 runtime mechanism. **Adopt existing runtime** requires an integrated and tested implementation; **Keep model-local** requires identical-input A/B evidence; **Needs framework change** requires a documented API gap and a follow-up issue. The selected final backend retains fixed-noise parity and the eager fallback. Report its latency and memory difference from R0 on the same data-center GPU, and justify any reproducible regression against the maintainability benefit. Unsupported reuse claims without code or profiling are not accepted.

### R2 — B>1 formed-batch execution

**Problem and deliverable.** The optimized Pi runtime is BS=1-oriented. Extend processor, pipeline, model, buffers, graph selection, and selected kernels to execute an already formed batch at B=1/2/4/8/16/32. This task delivers both B>1 functionality and throughput results; server-side request formation is not part of R2.

**Suggested approach.** Extend the R1 backend instead of maintaining a second batch path. Batch only compatible requests (view layout, fixed 200-token text layout, horizon, step count, dtype, and backend), while retaining per-row masks and state. Tune/capture only the declared batch buckets.

**Acceptance.** Batched fixed-noise outputs match independent R1 B=1 runs at `atol=1e-4`, including mixed per-row inputs, with no cross-request contamination. On the selected data-center GPU, report B=1/2/4/8/16/32 latency, throughput, scaling efficiency, B=1 regression, and memory; compare every point with sequential R1 B=1 execution and measured run-to-run variation. If a point cannot run, report the exact memory/capability limit rather than omitting it. B>1 functionality may be accepted without a throughput gain, but the result must not claim batching as a performance improvement.

### D0 — Scheduler-visible denoise steps and Continuous Batching

**Problem and deliverable.** Pi currently completes all denoise steps inside one model call, so the scheduler cannot admit or retire requests between steps. Prototype Pi on the existing step-execution contract and evaluate whether step-level Continuous Batching benefits fixed 10-step serving. Keep the monolithic path behind a flag/fallback.

**Suggested approach.** Reuse `STEP_BATCH`, `StepScheduler`, and dynamic `InputBatch`. Keep per-request prefix KV, latent, timestep, noise, and completion/cancellation state on device. A scheduler-visible denoise step is a logical scheduling boundary, not necessarily a separate process or physical pipeline stage. See the fixed-work proposal in [#6524](https://github.com/vllm-project/vllm-omni/issues/6524#issuecomment-5392110789).

**Acceptance.** Fixed-noise results match the monolithic R1 backend; dynamic admission, retirement, and cancellation do not corrupt other requests. On the selected data-center GPU, report N=1/2/4/8/16/32 burst and staggered arrivals, effective batch size, req/s, p50/p95/p99, scheduler overhead, utilization, and memory. Compare the latency-throughput Pareto frontier with serial R1 and, when available, R2 formed batching. Recommend adoption only if step-level Continuous Batching improves at least one tested operating point beyond measured run-to-run variation without an unexplained correctness, tail-latency, starvation, or memory regression; otherwise keep it experimental or reject it. A reproducible negative result with an Adopt/Experimental/Reject decision also completes this exploration.

## Out of scope and related work

- Post-training support will be tracked after the serving/runtime baseline is stable.
- MEM, RTC, cross-session prefix caching, timestep reduction, sequence parallelism, and Orin/Thor deployment are not included here.
- Related: [#6524](https://github.com/vllm-project/vllm-omni/issues/6524), Pi0 [#4222](https://github.com/vllm-project/vllm-omni/pull/4222), Pi0.5 draft [#4419](https://github.com/vllm-project/vllm-omni/pull/4419), and the Robot Policy Serving Contract [#6069](https://github.com/vllm-project/vllm-omni/issues/6069).

To claim a task, reply with its ID and planned scope. Add the owner and child issue/PR to the table; update the dependency status when a task lands.

**Comment by avraichur96** [comment]
Hi!  @tzhouam @gcanlin 
I'd like to contribute to this effort.

I'm interested in taking F2 (Pi0.5 parity and serving tests), particularly the reference harness, unit tests, and OpenPI end-to-end validation.

I'd also be happy to take ownership of a well-scoped implementation component from F1 (or another component you think is appropriate), so that the implementation and validation can progress in parallel.

If you prefer, I'm happy to work through child issues and coordinate closely with the F1 owner..

Thanks!

**Comment by VLa1111** [comment]
Hi @tzhouam @gcanlin — I'd like to claim S0 (four-layer timing harness and engine-path breakdown).

 **Plan**
  1. Reproducible four-layer timing harness — model-only, in-process pipeline, engine path, server E2E — matching the BS=1 breakdown already reported in the profiling comment on #6524. Built directly on the merged Pi0 path first (as the tracker allows), with a documented way to re-run it on Pi0.5 once F1 lands.
  2. Engine-path breakdown — separate orchestrator, queue/IPC, request construction, and worker handoff segments on top of the harness, so the engine path (currently the largest of the four layers with no internal split) is segmented.
  3. Outcome — either close the "rewrite the scheduler" direction with the measured non-model overhead, or open concrete optimization issues backed by the segmented data.

**Comment by princepride** [comment]
😅, I have some questions and I think you need give more detail description of each task:
> Four-layer timing harness and engine-path breakdown

Seems you need add more details of this task

> Pi0.5 parity and serving tests

Each unit test should implement with the feature

> Pi0.5 batch execution correctness and throughput

Seems this is also a unit test task

**Comment by BruceLoveDecimal** [comment]
Hi, I'm interested in Pi0.5 and have been following your integration PR. From the numbers in the PR, BS=1 looks launch-bound . Can I contribute CUDA Graph and torch.compile performance work stacked on top of 
#6950?

**Comment by shijiew555** [comment]
Hi, I am interested in taking task R0. Here is my plan:

1. Port useful kernel optimizations from #4419 incrementally. First port and retain optimized/unoptimized versions for Triton kernels, weight packing, and static buffers components. Decide whether to keep optimization by measuring each component.
2. Implement bounded, prewarmed CUDA Graph families for image encoding, VLM prefix, and single-step denoising, then compare against the monolithic prefix/denoise graph.
3. Benchmark results: validate fixed-noise parity across 1/2/3 views and varied prompt lengths, and report latency, initialization costs, and peak memory on a data-center GPU and consumer-GPU baselines.

**Comment by chenchaoxu7575** [comment]
Thanks @avraichur96 and @VLa1111, and sorry for the delayed follow-up. I substantially revised this RFC last week, including the task definitions, dependencies, benchmark contract, and acceptance criteria, so the current scopes differ from when you originally commented.

@avraichur96, parity and serving tests are now part of F1 and are covered by #6950. F2 now focuses on the shared Pi0/Pi0.5 model and runtime structure. Could you please reread the current F2 definition and confirm whether you are still interested in it, or select another task?

@VLa1111, could you please reread the current S0 definition and confirm whether you still want to take it? If so, please update your plan where needed based on the revised scope and acceptance criteria.

**Comment by chenchaoxu7575** [comment]
Thanks @shijiew555 and @BruceLoveDecimal. Since the CUDA Graph, torch.compile, Triton kernel, weight-packing, and static-buffer work is tightly coupled, I would prefer to keep R0 as one unified effort.

@shijiew555, would you be willing to own and coordinate R0, and open a child issue with the overall implementation and benchmark plan?

@BruceLoveDecimal, your earlier CUDA Graph and torch.compile proposal fits directly into this work. Would you like to join @shijiew555 and collaborate through the same child issue and integration path?

R0 will start after #6950 lands.

**Comment by avraichur96** [comment]
Thanks for the update, @chenchaoxu7575. I’ve reread the revised F2 definition and I’m still interested in taking it.

I understand that F2 now concerns the shared Pi0/Pi0.5 model and runtime structure—not parity-test implementation. I’ll begin with the Pi0/Pi0.5 difference table and propose a sharing boundary for review.

I’ll preserve both eager paths and the existing unit, parity, and OpenPI serving behaviour, and report any BS=1 latency change.

Please assign this to me. 

Thanks!

**Comment by shijiew555** [comment]
Thank you for the update @chenchaoxu7575. I am able to own and coordinate R0. I'll open a child RFC issue to formulate the division of work, test plans, and see how we can work together.

Thank you!

**Comment by chenchaoxu7575** [comment]
Thanks @avraichur96. F2 is yours. You can start with the Pi0/Pi0.5 difference table and proposed sharing boundary against the current #6950 branch. Please open a child issue for the proposal, and start the implementation after #6950 lands. I’ll update the tracker with you as the F2 owner.

**Comment by zhangj1an** [comment]
I can help review @shijiew555‘s R0, also implement R1 after R0 is merged. 🙋

#### Some other context info 

I tested pi0.5 on openpi ( pi0.5's official infra framework, checkpoint  `pi05_libero`) Pytorch on rtx5090 for batch size = 1, e2e client call p99 takes 32.30ms. After applying cuda graph replay and also nvfp4 quantisation, it drops to 25.55ms.  Quantisation decreases task success rate from 97.25% to 96.75%. 

The main reason for speedup is openpi predicts only next 10 (prev 50) actions steps per action chunk. And robot only executes next 5 (prev 10) steps then replan. As my company's robot's control frequency is 100hz, maybe inference speed of >20hz (<50ms) is already enough. I will try openpi version on real robot these 2 weeks. If it works I will reduce these 2 params accordingly too. 

I have also tried adding spec decode to pi0.5 following https://arxiv.org/abs/2605.13778 but the accuracy dropped from 96.75% to 91.6% on `libero_spatial` tasks, so it is not usable for now.

**Comment by avraichur96** [comment]
@chenchaoxu7575  

Added child for F2 issue. #7709 

Thanks!

**Comment by shijiew555** [comment]
@chenchaoxu7575 @zhangj1an @BruceLoveDecimal 
Added a child issue for R0: #7754 

Any suggestions are welcome. Thanks!
